### ISSUE TITLE
Recursive resolving parameter types in Arbitraries.forType

### ISSUE THREAD

**Issue by hohserg1** [jlink/jqwik#191]
## Testing Problem
Sometimes need to create randomized values of some Class<?>.
As example, I need it in tests of my annotation processor.

Follow code should work fine:
```java
public static class A {}

public static class B {}

public static class C {
        final A a1;
        final B b1;

        public C(A a1, B b1) {
            this.a1 = a1;
            this.b1 = b1;
        }
}

@Test
void doTest() {
        println( Arbitraries.forType(A.class).sample() ); //fine
        println( Arbitraries.forType(B.class).sample() ); //fine
        println( Arbitraries.forType(C.class).sample() ); //bad
}
```
But it crashes with `net.jqwik.api.CannotFindArbitraryException: Cannot find an Arbitrary for Parameter of type [A]`.
Although Arbitrary for A.class and B.class work fine

## Suggested Solution
Arbitraries.forType should resolve types of constructor parameters recursively.
Logic is something like:
```
Arbitraries.forType(C.class) ->
    //constructor of C need instances of A and B, ok
    Arbitraries.forType(A.class) ->
        //constructor of A no need instances of unknown types
        //build
    Arbitraries.forType(B.class)
        //constructor of B no need instances of unknown types
        //build
    //build
```

## Discussion
advantages:
+ forType will be more useful
+ not conflict with existing features

disadvantages:
+ need to implement

related with https://github.com/jlink/jqwik/issues/168

**Comment by jlink** [comment]
@hohserg1 Thanks for the suggestion. 

It's been on the longlist for sometime, having an official issue is all the better.

Here's what I've been planning:
- The standard behaviour remains the same
- An optional parameter will be introduced in `Arbitaries.forType(Class<T> targetType, int depth)`
- An optional parameter will be introduced for the annotation `@UseType(depth = 1)`

Would that suit your needs?

One point is: How should types be handled that can be resolved by default or explicitly registered arbitrary providers?
Should those providers take precedence? Should resolution by type have priority? Should that be configurable?

**Comment by hohserg1** [comment]
> Would that suit your needs?

Yeah, that's better, more control - better. 
Also, I think that need to keep overloading with only one argument for save compatibility:
```java
Arbitaries.forType(Class<T> targetType) -> Arbitaries.forType(targetType, 0)
```


> How should types be handled that can be resolved by default or explicitly registered arbitrary providers?

Two strategies come to mind:
+ Use already available arbitraries
+ Anyway create new arbitraries for each type (useful if available arbitrary is more bounded)
How about to add third argument which control this cases?

**Comment by jlink** [comment]
> Two strategies come to mind:
> 
> * Use already available arbitraries
> * Anyway create new arbitraries for each type (useful if available arbitrary is more bounded)
> 
> How about to add third argument which control this cases?

Having thought about it for a while I tend to go with the first strategy and no 3rd parameter. If more nuanced control over nested types is required domains can be used. 

BTW, domains are also my first choice when I want to handle your use case now.

**Comment by hohserg1** [comment]
Ok, extracting abstraction layer for it look better

**Comment by jlink** [comment]
Working on it.

First cut will probably be simpler and just offer `TypeArbitrary.allowRecursion()`.

**Comment by jlink** [comment]
Parameters annotated with `@UseType` now resolve recursively by default.

**Comment by jlink** [comment]
Functionality available in 1.6.1-SNAPSHOT.

See: https://jqwik.net/docs/snapshot/user-guide.html#generation-from-a-types-interface
